src/nexusrpc/handler/_serializer.py

```diff
@@ -0,0 +1,99 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import (
+    Any,
+    AsyncIterable,
+    Mapping,
+    Optional,
+    Protocol,
+    Type,
+)
+
+
+@dataclass
+class Content:
+    """
+    A container for a map of headers and a byte array of data.
+
+    It is used by the SDK's Serializer interface implementations.
+    """
+
+    headers: Mapping[str, str]
+    """
+    Header that should include information on how to deserialize this content.
+    Headers constructed by the framework always have lower case keys.
+    User provided keys are treated case-insensitively.
+    """
+
+    data: Optional[bytes] = None
+    """Request or response data. May be undefined for empty data."""
+
+
+class Serializer(Protocol):
+    """
+    Serializer is used by the framework to serialize/deserialize input and output.
+    """
+
+    # TODO(preview): support non-async def
+
+    async def serialize(self, value: Any) -> Content:
+        """Serialize encodes a value into a Content."""
+        ...
+
+    # TODO(prerelease): does None work as the sentinel type here, meaning do not attempt
+    # type conversion, despite the fact that Python treats None as a valid type?
+    async def deserialize(
+        self, content: Content, as_type: Optional[Type[Any]] = None
+    ) -> Any:
+        """Deserialize decodes a Content into a value.
+
+        Args:
+            content: The content to deserialize.
+            as_type: The type to convert the result of deserialization into.
+                     Do not attempt type conversion if this is None.
+        """
+        ...
+
+
+class LazyValue:
+    """
+    A container for a value encoded in an underlying stream.
+    It is used to stream inputs and outputs in the various client and server APIs.
+    """
+
+    def __init__(
+        self,
+        serializer: Serializer,
+        headers: Mapping[str, str],
+        stream: Optional[AsyncIterable[bytes]] = None,
```

Comment:
I wonder if this is more of an `asyncio.StreamReader`. Though I am not sure that we should force people to use async.